### PR TITLE
feat: fully configurable enemy weapons

### PR DESCRIPTION
**Pull request by ghost**
image

image

image

### CODE REVIEW

**Comment by Jambu1989** [comment]
I freaking hate git

**Comment by ghost** [comment]
This should be ready to finally review now. Had to max some emergency sanity changes. :|

File: src/base/qst.cpp [hunk 1]
```diff
@@ -9833,6 +9834,87 @@ static void guys_update_firesfx(guydata& tempguy)
 			}
 		}
 	}
+	
+	tempguy.wunblockable = 0;
```
Comment (connorjclark) [inline src/base/qst.cpp]: how come all this weapon stuff is repeated in the `guy_update_firesfx` method?

File: src/zc/guys.h [hunk 2]
```diff
@@ -115,6 +115,15 @@ class enemy : public sprite
 	
 	bool didScriptThisFrame; //Hack for enemies which are spawned by shit like scripted load enemies so they don't run their script a second time in a frame.
 	
+	byte wunblockable;
+	move_flags wmoveflags;
+	int32_t weapoverrideFLAGS;
```
Comment (connorjclark) [inline src/zc/guys.h:120]: flag variables should be unsigned

File: src/zc/weapons.cpp [hunk 3]
```diff
@@ -4337,10 +4411,9 @@ bool weapon::animate(int32_t index)
 				sfx(itemsbuf[parentitem].usesound,pan(int32_t(x)),true,false);
 			}
 			if(runscript_do_earlyret(run_script(MODE_NORMAL))) return false;
+
 		}
 		
-		break;
```
Comment (connorjclark) [inline src/zc/weapons.cpp:4342]: Why did this break go away?

File: src/zc/ffscript.cpp [hunk 4]
```diff
@@ -10121,7 +10121,7 @@ int32_t get_register(int32_t arg)
 				Z_scripterrlog("Invalid Sprite ID passed to npcdata->Attributes[]: %d\n", (ri->npcdataref*10000)); 
 				ret = -10000;
 			}
-			else if ( indx < 0 || indx > MAX_NPC_ATTRIBUTES )
+			else if ( indx < 0 || indx >= MAX_NPC_ATTRIBUTES )
```
Comment (connorjclark) [inline src/zc/ffscript.cpp:10124]: We should cherry pick these fixes into 2.55, but this larger change won't be going to 2.55 and is unrelated. I'll split it out for you.

File: src/base/qst.cpp [hunk 5]
```diff
@@ -9833,6 +9834,87 @@ static void guys_update_firesfx(guydata& tempguy)
...
+		break;
+	}
+	for (int q=0; q < WPNSPR_MAX; ++q)
+	{
+		tempguy.burnsprs[q] = 0;
+		tempguy.light_rads[q] = 0;
+	}
+}
+
+static void guy_update_weaponflags(guydata& tempguy)
```
Comment (connorjclark) [inline src/base/qst.cpp:9837]: so the "update firesfx" method was so we can update all the guys and not need to introduce a QR for any compat

can't we do the same here, in order to remove the QR? in otherwords, for wunblockable and wmoveflags and wstep, set those values based on the guy's weapon.

(for weaponoverrideFLAGS and guy_burning_sprites, that is handled by just setting weapoverrideFLAGS and guy_burning_sprites to 0 )

File: src/zc/ffscript.cpp (hunk 4 above)
Comment (connorjclark) [reply src/zc/ffscript.cpp:10124]: ah this is just b/c you updated the constant to be more reasonable. thought this was a bug fix, but it's just a refactor. still...i'll just split it out anyhow

File: src/base/qst.cpp (hunk 1 above)
Comment (ghost) [reply src/base/qst.cpp]: fixed

File: src/base/qst.cpp (hunk 5 above)
Comment (ghost) [reply src/base/qst.cpp:9837]: that is legit what happens in update_Weaponflags()

Comment (connorjclark) [reply src/base/qst.cpp:9837]: ah great, then I'm confused why a QR is needed if these values get set there?

File: src/zc/weapons.cpp (hunk 3 above)
Comment (ghost) [reply src/zc/weapons.cpp:4342]: fixed: meant to move it into the curlybraces above!

File: src/base/qst.cpp (hunk 5 above)
Comment (ghost) [reply src/base/qst.cpp:9837]: I only did it cause of this nonsense.
We can do a replay check thou.
https://discord.com/channels/876899628556091432/1280357903265693798/1280357903265693798

Comment (Jambu1989) [reply src/base/qst.cpp:9839]: oh and echnically we need the method anyways cause of init_guys

File: src/zc/guys.h (hunk 2 above)
Comment (ghost) [reply src/zc/guys.h:120]: can i refactor this separately since this is more then just here?

Comment (connorjclark) [reply src/zc/guys.h:120]: sure

File: src/base/qst.cpp (hunk 5 above)
Comment (ghost) [reply src/base/qst.cpp:9837]: fixed

**Review by connorjclark — APPROVED** [review]
perfect

**Pull request merged**